rfcs/20221124-monorepo.md

```diff
@@ -0,0 +1,179 @@
+
+1. Milestones that can span across multiple layers of the stack are only possible in GitHub. This is a limitation imposed by GitHub and there is no workaround for this.
+
+The [integration](#step-4-integration) section in the latter part of the document describes more interesting outcomes made possible by the monorepo. They may or may not be exclusive to monorepos but they're surely made easier by it.
+
+## Migration path
+
+First we will merge the API and the frontend. This decision was made for the following reasons.
+
+1. API and frontend are tightly linked. The frontend is a direct consumer of what the API produces.
+
+1. The API and frontend form the "service" side of Openverse that directly faces the users (both API consumers and Search engine users).
+
+1. The frontend uses ECS deployments and the API is well on the same track. This makes it possible for them to share some deployment code.
+
+1. To the RFC author, the API and frontend are very familiar so merging them would be easier. Adding a third component would make the task daunting.
+
+1. Merging incurs a productivity hit for the initial transition. So merging everything in one swoop is not ideal.
+
+1. The API’s comprehensive tooling for developer documentation can benefit frontend devs and create a unified docs site for contributors.
+
+1. The API is already organised by stack folders so the `frontend/` directory will fit right in with the others like `api/` and `ingestion_server/`.
+
+1. The API and frontend share identical tooling for Git hooks, linting and formatting. We will fight our tools less and encounter minimal friction.
+
+   - In fact, we employ a number of hacks to install and configure pre-commit for the frontend. Merging it with the API eliminates the need for such hacks.
+
+1. The entire system can be integration tested during releases. The real API, populated with test data, can even replace the Talkback server.
+
+The `WordPress/openverse-api` repo will absorb the `WordPress/openverse-frontend` repo. The `WordPress/openverse-catalog` will also be merged, _later_.
+
+### Reference
+
+I'm following the steps listed below in a fork at [@dhruvkb/monopenverse](https://github.com/dhruvkb/monopenverse/). You can refer to the fork, but note that it is a comes from a place of haste and has not been treated with the same level of love and care that the final treatment will receive.
+
+### Step 0: Prerequisites
+
+#### Get the timing right
+
+The first step will be to release the frontend, call a code freeze and pause work on it. This is to prevent the frontend repo from continuing to drift as we merge a snapshot of it with the API.
+
+This can prove difficult given how productive our team is, so we will need to channel this productivity towards the catalog in the meantime. I can foresee the end-to-end migration taking one week (ideal scenario) to one fortnight (worst case scenario).
```

Comment:
> I can foresee the end-to-end migration taking one week (ideal scenario) to one fortnight (worst case scenario).

Could you unpack this time estimate a bit more for me? I am curious because I could see these tasks taking an hour or so and being treated like a deploy, where a few folks work on it together synchronously. I also think that would have benefit of minimizing the disruption to contributors. Are there long-running tasks I'm not considering?

Reply:
The missing implementation details (such as writing new documentation, updating references) are what will take the maximum time in the 1-2 week time range I proposed for this plan.

The technical merge is a and getting the functionality working is a day or two. The remaining 3 days are for identifying these issues and fixing them. I kept 5 more days as a backup in case something unforeseen comes up but given we do sufficient planning beforehand, we will not need the second week (underpromise, overdeliver 😉).

It might be hard to make a draft PR with the changes (because of the merge conflicts) but one thing we could do is start a new branch, do the merge on it and eventually replace `main` with it. That seems like a great way to see everything in action but a lot of (branch-name dependent actions will be affected when we promote the other branch).

In any case, the frontend cannot be merged via the GitHub UI, it will have be CLI. In all my usage of GitHub I don't think it's possible to make a PR from any repo to another. The GitHub UI only provides merge functionality from a fork to the upstream which is not this case.

Reply:
I'd still really like to see a breakdown of the tasks involved in the merge, their execution order, and a time estimate for each step. We could do that when we write an implementation plan for this RFC. Two days maximum for a code freeze isn't too disruptive but definitely a concern in the case of high priority or critical bugs we might identify, even with a bug bash and small code freeze beforehand.